src/Target.cpp

```diff
@@ -390,34 +333,105 @@ bool Target::merge_string(const std::string &target) {
     }
 
     // If arch is PNaCl, require explicit setting of os and bits as well.
-    if (arch_specified && arch == Target::PNaCl) {
-        if (!os_specified || os != Target::NaCl) {
+    if (arch_specified && t.arch == Target::PNaCl) {
+        if (!os_specified || t.os != Target::NaCl) {
             return false;
         }
-        if (!bits_specified || bits != 32) {
+        if (!bits_specified || t.bits != 32) {
             return false;
         }
     }
 
     return true;
 }
 
+void bad_target_string(const std::string &target) {
+    const char *separator = "";
+    std::string architectures;
+    for (const auto &arch_entry : arch_name_map) {
+        architectures += separator + arch_entry.first;
+        separator = ", ";
+    }
+    separator = "";
+    std::string oses;
+    for (auto os_entry : os_name_map) {
+        oses += separator + os_entry.first;
+        separator = ", ";
+    }
+    separator = "";
+    // Format the features to go one feature over 70 characters per line,
+    // assume the first line starts with "Features are ".
+    int line_char_start = -(int)sizeof("Features are");
+    std::string features;
+    for (auto feature_entry : feature_name_map) {
+        features += separator + feature_entry.first;
+        if (features.length() - line_char_start > 70) {
+            separator = "\n";
+            line_char_start = features.length();
+        } else {
+            separator = ", ";
+        }
+    }
+    user_error << "Did not understand HL_TARGET=" << target << "\n"
+               << "Expected format is arch-os-feature1-feature2-...\n"
+               << "Where arch is " << architectures << " .\n"
+               << "Os is " << oses << " .\n"
+               << "If arch or os are omitted, they default to the host.\n"
+               << "Features are " << features << " .\n"
+               << "HL_TARGET can also begin with \"host\", which sets the "
+               << "host's architecture, os, and feature set, with the "
+               << "exception of the GPU runtimes, which default to off.\n"
+               << "On this platform, the host target is: " << get_host_target().to_string() << "\n";
+}
+
+}
+
+Target::Target(const std::string &target) {
+    Target host = get_host_target();
+
+    if (target.empty()) {
+        // If nothing is specified, use the full host target.
+        *this = host;
+    } else {
+
+        // Default to the host OS and architecture in case of partially
+        // specified targets (e.g. x86-64-cuda doesn't specify the OS, so
+        // use the host OS).
+        os = host.os;
+        arch = host.arch;
+        bits = host.bits;
+
+        if (!merge_string(*this, target)) {
+            bad_target_string(target);
+        }
+    }
+}
+
+Target::Target(const char *s) {
+    *this = Target(std::string(s));
+}
+
+bool Target::validate_target_string(const std::string &s) {
+    Target t;
+    return merge_string(t, s);
```

Comment:
It seems a little wasteful to throw away the result: if you need to check and fail in a way other than user_error, you parse twice. Not sure if this matters.